crates/red_knot_python_semantic/src/types/call.rs

```diff
@@ -140,6 +152,32 @@ impl<'db> CallError<'db> {
         self.return_type(db).unwrap_or(Type::unknown())
     }
 
+    /// Sets the return type of the call.
+    pub(super) fn set_return_type(&mut self, return_ty: Type<'db>) {
+        match self {
+            CallError::NotCallable { .. } => {}
+            CallError::Union(UnionCallError {
+                bindings, errors, ..
+            }) => {
+                // We need to recursively replace the return type of each binding
+                // inside the union, and traverse nested errors.
+                for binding in bindings {
+                    binding.set_return_type(return_ty);
+                }
+
+                for error in errors {
+                    error.set_return_type(return_ty);
+                }
+            }
+            CallError::PossiblyUnboundDunderCall { outcome, .. } => {
+                outcome.set_return_type(return_ty);
+            }
+            CallError::BindingError { binding } => {
+                binding.set_return_type(return_ty);
+            }
+        }
```

Comment:
This is a little ugly, and feels like throwing away information. I'd almost rather just have an optional `override_return_ty` field.